src/Storages/StorageDelta.cpp

```diff
@@ -0,0 +1,357 @@
+}
+
+std::vector<String> DeltaLakeMetadata::ListCurrentFiles() &&
+{
+    std::vector<String> keys;
+    keys.reserve(file_update_time.size());
+
+    for (auto && [k, _] : file_update_time)
+    {
+        keys.push_back(k);
+    }
+
+    return keys;
+}
+
+JsonMetadataGetter::JsonMetadataGetter(StorageS3::S3Configuration & configuration_, const String & table_path_, Poco::Logger * log_)
+    : base_configuration(configuration_), table_path(table_path_), log(log_)
+{
+    Init();
+}
+
+void JsonMetadataGetter::Init()
+{
+    auto keys = getJsonLogFiles();
+
+    // read data from every json log file
+    for (const String & key : keys)
+    {
+        auto buf = createS3ReadBuffer(key);
+
+        while (!buf->eof())
+        {
+            String json_str = readJSONStringFromBuffer(buf);
+            
+            if (json_str.empty()) {
+                continue;
+            }
+
+            const JSON json(json_str);
+
+            handleJSON(json);
+        }
+    }
+}
+
+std::vector<String> JsonMetadataGetter::getJsonLogFiles()
+{
+    std::vector<String> keys;
+
+    const auto & client = base_configuration.client;
+
+    Aws::S3::Model::ListObjectsV2Request request;
+    Aws::S3::Model::ListObjectsV2Outcome outcome;
+
+    bool is_finished{false};
+    const auto bucket{base_configuration.uri.bucket};
+
+    request.SetBucket(bucket);
+    request.SetPrefix(std::filesystem::path(table_path) / "_delta_log");
+
+    while (!is_finished)
+    {
+        outcome = client->ListObjectsV2(request);
+        if (!outcome.IsSuccess())
+            throw Exception(
+                ErrorCodes::S3_ERROR,
+                "Could not list objects in bucket {} with key {}, S3 exception: {}, message: {}",
+                quoteString(bucket),
+                quoteString(table_path),
+                backQuote(outcome.GetError().GetExceptionName()),
+                quoteString(outcome.GetError().GetMessage()));
+
+        const auto & result_batch = outcome.GetResult().GetContents();
+        for (const auto & obj : result_batch)
+        {
+            const auto & filename = obj.GetKey();
+
+            if (std::filesystem::path(filename).extension() == ".json")
+                keys.push_back(filename);
+        }
+
+        request.SetContinuationToken(outcome.GetResult().GetNextContinuationToken());
+        is_finished = !outcome.GetResult().GetIsTruncated();
+    }
+
+    return keys;
+}
+
+std::shared_ptr<ReadBuffer> JsonMetadataGetter::createS3ReadBuffer(const String & key)
+{
+    // TBD: add parallel downloads
+    return std::make_shared<ReadBufferFromS3>(
+        base_configuration.client,
+        base_configuration.uri.bucket,
+        key,
+        base_configuration.uri.version_id,
+        /* max single read retries */ 10,
+        ReadSettings{});
+}
+
+String JsonMetadataGetter::readJSONStringFromBuffer(std::shared_ptr<ReadBuffer> buf) {
```

Comment:
Looks like you can use https://github.com/ClickHouse/ClickHouse/blob/d59880298868d84c5c6bc38753c9308db6b5db88/src/IO/ReadHelpers.h#L541
instead.

Reply:
Used readJSONObjectPossiblyInvalid as it is more suitable in this case.